receiver/signalfxreceiver/signalfxv2_to_metricdata.go

```diff
@@ -17,214 +17,190 @@ package signalfxreceiver
+		return errSFxNoDatumValue
 	}
 
-	p := &metricspb.Point{
-		Timestamp: convertTimestamp(sfxDataPoint.GetTimestamp()),
-	}
+	dps.Resize(1)
+	dp := dps.At(0)
 
-	switch {
-	case sfxDataPoint.Value.IntValue != nil:
-		mismatch := expectedMetricType != metricspb.MetricDescriptor_CUMULATIVE_INT64 &&
-			expectedMetricType != metricspb.MetricDescriptor_GAUGE_INT64
-		if mismatch {
-			return nil, errSFxUnexpectedInt64DatumType
-		}
-		p.Value = &metricspb.Point_Int64Value{Int64Value: *sfxDataPoint.Value.IntValue}
+	dp.SetTimestamp(dpTimestamp(sfxDataPoint))
+	dp.SetValue(*sfxDataPoint.Value.IntValue)
+	fillInLabels(sfxDataPoint.Dimensions, dp.LabelsMap())
 
-	case sfxDataPoint.Value.DoubleValue != nil:
-		mismatch := expectedMetricType != metricspb.MetricDescriptor_CUMULATIVE_DOUBLE &&
-			expectedMetricType != metricspb.MetricDescriptor_GAUGE_DOUBLE
-		if mismatch {
-			return nil, errSFxUnexpectedFloat64DatumType
-		}
-		p.Value = &metricspb.Point_DoubleValue{DoubleValue: *sfxDataPoint.Value.DoubleValue}
+	return nil
+}
 
-	case sfxDataPoint.Value.StrValue != nil:
-		if expectedMetricType != metricspb.MetricDescriptor_GAUGE_DOUBLE {
-			return nil, errSFxUnexpectedStringDatumType
-		}
-		dbl, err := strconv.ParseFloat(*sfxDataPoint.Value.StrValue, 64)
-		if err != nil {
-			return nil, errSFxStringDatumNotNumber
-		}
-		p.Value = &metricspb.Point_DoubleValue{DoubleValue: dbl}
+func fillDoubleDataPoint(sfxDataPoint *sfxpb.DataPoint, dps pdata.DoubleDataPointSlice) error {
+	if sfxDataPoint.Value.DoubleValue == nil {
+		return errSFxNoDatumValue
 	}
 
-	return p, nil
-}
+	dps.Resize(1)
+	dp := dps.At(0)
 
-func convertTimestamp(msec int64) *timestamppb.Timestamp {
-	if msec == 0 {
-		return nil
-	}
+	dp.SetTimestamp(dpTimestamp(sfxDataPoint))
+	dp.SetValue(*sfxDataPoint.Value.DoubleValue)
+	fillInLabels(sfxDataPoint.Dimensions, dp.LabelsMap())
 
-	ts := &timestamppb.Timestamp{
-		Seconds: msec / 1e3,
-		Nanos:   int32(msec%1e3) * 1e6,
-	}
-	return ts
-}
+	return nil
 
-func buildDescriptor(
-	sfxDataPoint *sfxpb.DataPoint,
-	labelKeys []*metricspb.LabelKey,
-	metricType metricspb.MetricDescriptor_Type,
-) *metricspb.MetricDescriptor {
-
-	// TODO: Evaluate performance impact with different datasets to see if it
-	//  is worth to cache these.
-	descriptor := &metricspb.MetricDescriptor{
-		Name: sfxDataPoint.GetMetric(),
-		// Description: no value to go here
-		// Unit:        no value to go here
-		Type:      metricType,
-		LabelKeys: labelKeys,
-	}
+}
 
-	return descriptor
+func dpTimestamp(dp *sfxpb.DataPoint) pdata.TimestampUnixNano {
+	return pdata.TimestampUnixNano(dp.GetTimestamp() * 1e6)
 }
 
-func buildLabelKeysAndValues(
+func fillInLabels(
 	dimensions []*sfxpb.Dimension,
-) ([]*metricspb.LabelKey, []*metricspb.LabelValue) {
-	keys := make([]*metricspb.LabelKey, 0, len(dimensions))
-	values := make([]*metricspb.LabelValue, 0, len(dimensions))
+	labels pdata.StringMap,
+) {
 	for _, dim := range dimensions {
```

Comment:
Call `lables.InitEmptyWithCapacity(len(dimensions))` to pre-allocate the memory.

Reply:
Done.